### PR TITLE
internal data structure rework

### PR DESCRIPTION
**Pull request by xofyarg**
This commit includes several changes:
- Use Bytes for parsing to reduce memory copy.
- Remove unnecessary async code.
- Add a client example.
- Make RNG parameter explicit, can be used in env which doesn't
  support regular RNG syscall.
- Fix panics while parsing unexpect data.
- Use concrete error types.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -10,20 +10,25 @@ keywords = ["odoh", "protocols", "dns", "doh", "privacy"]
...
+clap = "3.0.0-beta.2"
 criterion = "0.3"
+domain = "0.5"
+env_logger = "0.8"
 hex = "0.4"
+log = "0.4"
+reqwest = "0.11"
 serde_json = "1.0"
-tokio = { version = "0.2", features = [ "full" ] }
+tokio = { version = "1.4", features = [ "full" ] }
```
Comment (vavrusa) [inline Cargo.toml]: You can probably stick with `^1`, it shouldn't matter

File: benches/odoh_bench.rs [hunk 2]
```diff
@@ -1,61 +1,59 @@
...
+    let key_pair = ObliviousDoHKeyPair::new(&mut rng);
 
-fn generate_query_body() -> ObliviousDoHQueryBody {
-    ObliviousDoHQueryBody::new(
-        &hex::decode("5513010000010000000000000377777706676f6f676c6503636f6d00001c0001").unwrap(),
-        None,
-    )
-}
+    c.bench_function("step_encrypt_query", |b| {
+        b.iter(|| {
```
Comment (vavrusa) [inline benches/odoh_bench.rs:28]: You can move input generation to `b.iter_batched(|| input() |input| ... bench ...)` to avoid benchmarking that.

File: src/draft06.rs [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+    #[error("Unexpected internal error")]
+    Internal,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Serialize to IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
+pub trait Serialize {
+    /// Serialize the provided struct into the buf.
+    fn serialize(self, buf: &mut BytesMut) -> Result<()>;
```
Comment (vavrusa) [inline src/draft06.rs]: You can use `<B: BufMut>`, that should allow either storage implementation.

File: src/draft06.rs [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+}
+
+/// Deserialize from IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
+pub trait Deserialize {
+    /// Deserialize a struct from the buf.
+    fn deserialize(buf: &mut Bytes) -> Result<Self>
```
Comment (vavrusa) [inline src/draft06.rs]: Similarly to Serialize, `<B: Buf>`.

File: src/draft06.rs [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+    query: &ObliviousDoHMessagePlaintext,
+    response: &ObliviousDoHMessagePlaintext,
+    secret: OdohSecret,
+    response_nonce: ResponseNonce,
+) -> Result<ObliviousDoHMessage> {
+    let buf = compose(response)?;
+
+    let (key, nonce) = derive_secrets(secret, query, response_nonce)?;
+    let cipher = Aes128Gcm::new(GenericArray::from_slice(&key));
+    let mut data = buf.to_vec();
```
Comment (vavrusa) [inline src/draft06.rs]: This will clone the composed response again

File: benches/odoh_bench.rs (hunk 2 above)
Comment (xofyarg) [reply benches/odoh_bench.rs:28]: Practically the interfaces be called after data deserializationc, so putting it into the bench can give a general idea of total cost.

File: src/draft06.rs (hunk 5 above)
Comment (xofyarg) [reply src/draft06.rs]: Unfortunately, the crypto interfaces only accept `Vec<u8>`. The data copy can be avoided here but not everywhere.

Comment (vavrusa) [reply src/draft06.rs]: You can probably implement aead::Buffer for BytesMut, but then ObliviousDoHMessage would have to be defined for owned type (BufMut) for `encrypted_msg` field. This specialization would only work in ObliviousDoHMessage could be deserialized from BytesMut.

File: src/draft06.rs [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+
+        let kem_id = buf.get_u16();
+        let kdf_id = buf.get_u16();
+        let aead_id = buf.get_u16();
+
+        if kem_id != KEM_ID || kdf_id != KDF_ID || aead_id != AEAD_ID {
+            return Err(Error::InvalidParameter);
+        }
+
+        let public_key = read_lengthed(&mut buf)?;
```
Comment (chris-wood) [inline src/draft06.rs:320]: Should we also validate the public key length (32B)?

**Review by chris-wood — APPROVED** [review]

File: src/draft06.rs (hunk 6 above)
Comment (chris-wood) [reply src/draft06.rs:320]: @xofyarg can we please fix this in a separate followup PR?

File: examples/cli.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
+// A basic client example.
```
Comment (tanyav2) [inline examples/cli.rs:1]: This exists here: https://github.com/cloudflare/odoh-client-rs

File: src/draft06.rs [hunk 8]
```diff
@@ -0,0 +1,756 @@
+//! API for protocol functionality such as creating and parsing ODoH queries and responses.
```
Comment (tanyav2) [inline src/draft06.rs:1]: I would recommend not using new draft names as filenames because it makes figuring out what changed between two versions of the protocol much harder as Github doesn't allow you to diff between files with different names.

If it is really necessary to do so, it should be done in a different PR where there are no differences between the contents of the files, and it is purely a name change.

**Comment by tanyav2** [comment]
I believe the test using the `test-vectors.json` was removed. That's pretty critical to ensure interop. There's other differences in testing as well, in particular, the config test has some hairy things that need to be ensured, which is lacking in the new configs test. There's also assertions that need to be made about the generated ciphertext given a fixed nonce to ensure that the right hpke version is being used. I would highly recommend keeping all the tests from the previous version, and if changes to testing have to be made, they should be made in a different PR.

File: src/protocol.rs [hunk 9]
```diff
@@ -1,843 +0,0 @@
...
-macro_rules! impl_custom_serde {
-    ($name:ident) => {
-        impl Serialize for $name {}
-        impl<'de> Deserialize<'de> for $name {}
-    };
-}
-
-/// Serialize to IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
-pub trait Serialize {
-    fn to_bytes(&self) -> Result<Vec<u8>>
```
Comment (tanyav2) [inline src/protocol.rs:59]: If you're trying to not allocate memory here, you can simply do 
```
    fn to_bytes<B: std::io::Write>(&self, buf: &mut B) -> Result<()>
    where
        Self: SerdeSerialize,
    {
        let serialized = bincode2::config()
            .big_endian()
            .array_length(LengthOption::U16)
            .serialize_into(buf, &self)?;
        Ok(serialized)
    }
    ```
  instead of removing `bincode2` entirely. It's much more concise as well thanks to the macro.

File: src/draft06.rs [hunk 10]
```diff
@@ -0,0 +1,756 @@
...
+    AesGcm(#[from] aes_gcm::Error),
+
+    /// Unexpected internal error.
+    #[error("Unexpected internal error")]
+    Internal,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Serialize to IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
```
Comment (tanyav2) [inline src/draft06.rs:103]: Instead of doing all serializations by hand, you can just use https://docs.rs/bincode2/2.0.1/bincode2/fn.serialize_into.html if you want to pre allocate memory, otherwise it's adding a lot of unnecessary boilerplate. I added more detail here: https://github.com/cloudflare/odoh-rs/pull/10#discussion_r619556666

File: benches/odoh_bench.rs [hunk 11]
```diff
@@ -1,61 +1,66 @@
...
+use rand::rngs::StdRng;
+use rand::SeedableRng;
 
-fn generate_query_body() -> ObliviousDoHQueryBody {
-    ObliviousDoHQueryBody::new(
-        &hex::decode("5513010000010000000000000377777706676f6f676c6503636f6d00001c0001").unwrap(),
-        None,
-    )
-}
+pub fn bench_steps(c: &mut Criterion) {
```
Comment (tanyav2) [inline benches/odoh_bench.rs:6]: I ran benches to compare the previous code vs this code, and the previous code is an order of magnitude faster:

`parse_received_query    time:   [255.23 ns 256.12 ns 257.72 ns]`: This is equivalent to `step_decrypt_query` as it parses and decrypts the query. `step_decrypt_query      time:   [92.795 us 93.415 us 94.195 us] `.

`parse_received_response time:   [305.71 ns 317.56 ns 334.49 ns]`: This is equivalent to `step_decrypt_response` as it parses and decrypts the response. `step_decrypt_response   time:   [8.5891 us 8.6432 us 8.7128 us]`

File: src/draft06.rs [hunk 12]
```diff
@@ -0,0 +1,756 @@
...
+        // to the clients.
+        let public_key = key_pair.public().clone();
+        let client_configs: ObliviousDoHConfigs = vec![ObliviousDoHConfig::from(public_key)].into();
+        let client_configs_bytes = compose(&client_configs).unwrap().freeze();
+
+        // ... distributing client_configs_bytes ...
+
+        // Parse and extract first supported config from client configs on client side.
+        let client_configs: ObliviousDoHConfigs = parse(&mut client_configs_bytes.clone()).unwrap();
+        let client_config = client_configs.into_iter().next().unwrap();
```
Comment (tanyav2) [inline src/draft06.rs:726]: The client config should not be manually extracted, this should be done by the library. See example here: https://github.com/cloudflare/odoh-client-rs/blob/master/src/main.rs#L52

**Comment by tanyav2** [comment]
Overall, I would recommend keeping the original tests and the original filename, and using https://github.com/cloudflare/odoh-client-rs/ for the client implementation, and consider using the original ser/deser logic with https://docs.rs/bincode2/2.0.1/bincode2/fn.serialize_into.html. Other than that, this looks great, thanks for adding in the error struct!

File: src/draft06.rs (hunk 6 above)
Comment (xofyarg) [reply src/draft06.rs:320]: Sure. I'll submit another one later.

File: examples/cli.rs (hunk 7 above)
Comment (xofyarg) [reply examples/cli.rs:1]: This is just an example, not mean to replace the client program.

File: src/draft06.rs (hunk 8 above)
Comment (xofyarg) [reply src/draft06.rs:1]: My initial thought was, maybe we would have to support multiple draft version at the same time. I agree it will make tracking a bit hard. I can change the filename back to previous one.

File: src/draft06.rs (hunk 10 above)
Comment (xofyarg) [reply src/draft06.rs:103]: I'm not sure it's a good idea to use bincode for this task, reasons:
1. Using a encoding not designed for the protocol would tangle the internal data structure with binary representation, this makes it hard to change the structure later without extra attention. E.g. in the spec, `ObliviousDoHConfig` has a field `length` after `version`, but we cannot have it here as the next field is an array, which implies the length when parsed by bincode .
2. With its popularity in rust world, bincode is not so well known in other places. It's not a standard, lacks of a specification about how the serialization and deserialization  are done. Besides, we were using a even less common fork states as `A maintained fork of the now-defunct Bincode.` It feels like there's a coincidence that the library "happen" to work for this protocol.

Comment (tanyav2) [reply src/draft06.rs:103]: That's fair, no worries then.

File: src/draft06.rs (hunk 8 above)
Comment (tanyav2) [reply src/draft06.rs:1]: I don't think supporting multiple versions is very useful, especially since we can do 0-downtime version upgrades now.
